### DISCUSSION TITLE
Ignore objects in global environment

### DISCUSSION THREAD

**Discussion by edalfon** [ropensci/targets#701] [Help]
Is there any way to get `{target}`'s static code analysis to ignore specific objects in the global environment?

I see `{targets}` uses `codetools::findGlobals()`, which does not provide any filtering functionality, and I've quickly browsed targets's code and couldn't find anything that could help here.

This perhaps can have other uses, but the use case I have in mind is this:

In our pipeline, data wrangling occurs mostly on a database. So many targets at the beginning of the pipeline need a connection for the DB backend. We create the connection in the global environment (e.g. first line in `_targets.R`) so that all targets that need it can use it. But many downstream targets work only in R and they do not need the connection anymore.

In this context, it was also a good idea to only establish the connection when a target actually needs it. This is to play nice with the DB backend. It also saves some time, as you do not have to wait for the connection to do non-DB work (e.g. `tar_visnetwork` does not need access to the DB, so it is really a waste to connect to the DB every time we want to visualize the pipeline -which happens VERY frequently-). And perhaps most importantly, because it would not trigger an error if the DB is not available, so that you can work in the downstream targets that do not need the DB.

Our attempt to achieve this was R's lazy evaluation: using `base::delayedAssign` to create a promise for the connection, that would only be established when the value of this promise is requested by a target. But this approach clashes with `{target}`'s static code analysis, because the promise for the connection will be identified in the global environment and then hashed, thereby requesting its value and effectively establishing the connection.

Thus, it would be really helpful to just ignore the `con` in the global environment and do not hash it, so that it remains as a promise. Any ideas on how to do this or workarounds?

We tried a number of workarounds:
- The most natural way seems defining the connection within a function and not directly in the global env. This partially works, as the function gets hashed and the connection is only established when you actually call the function. It has downsides, though: it does not play so well with `{knitr}`'s sql engine. In addition, you would then need to manage the connections to prevent leakage and breaking connections limit (which took us to the next workaround), and, again, to try and play nice with the DB backend.
- Then we experimented with the `pool` package. Here again, it does not play nice with `{knitr}`'s sql engine (e.g. `Pool does not implement the dbSendQuery() method`).

**Comment by wlandau** [comment]
> Is there any way to get {target}'s static code analysis to ignore specific objects in the global environment?

In the target's command, you can reference the global object with `eval(quote(object_name))` instead of  just `object_name`.

> In our pipeline, data wrangling occurs mostly on a database. So many targets at the beginning of the pipeline need a connection for the DB backend. We create the connection in the global environment (e.g. first line in _targets.R) so that all targets that need it can use it.

That will work for `tar_make()`, but not for parallel processing with `tar_make_clustermq()` or `tar_make_future()` because database connection objects are [not exportable](https://cran.r-project.org/web/packages/future/vignettes/future-4-non-exportable-objects.html) to parallel workers. I recommend instead that you set up and tear down the connection object locally within each target. You could accomplish this with something like:

```r
# _targets.R
# ...
tarchetypes::tar_hook_before(
  target_list,
  hook = {
    connection <- dbConnect(...)
    on.exit(close(connection))
  }
)
```

Minimal example with the hook:

``` r
library(targets)
tar_script({
  library(tarchetypes)
  tar_hook_before(
    list(
      tar_target(x1, print("middle1")),
      tar_target(x2, print("middle2"))
    ),
    hook = {
      print("beginning")
      on.exit(print("end"))
    }
  )
})
tar_make()
#> • start target x1
#> [1] "beginning"
#> [1] "middle1"
#> [1] "end"
#> • built target x1
#> • start target x2
#> [1] "beginning"
#> [1] "middle2"
#> [1] "end"
#> • built target x2
#> • end pipeline
```

<sup>Created on 2021-11-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by edalfon** [reply]
Many thanks for your answer, and of course, for `{targets}`: such a great package!

> I recommend instead that you set up and tear down the connection object locally within each target. 

I guess that'd work. We were actually trying to avoid opening and closing the connection in each target. Instead, we opted for setting up the connection in a project-level `.Rprofile`. And even though it is not currently our use case, if at some point we wanted to use `tar_make_clustermq()` or `tar_make_future()`, we would just need to make sure all processes start with the same `.Rprofile`. An additional perk of setting up the connection in the `.Rprofile`: it is readily available for interactive work (as a promise). So you just need to fire up the R session and start interacting with the DB. 

But yeah, the hooks-based solution you recommend seems much cleaner/robust. Many thanks again.
